internal/charmstore/search.go

```diff
@@ -98,11 +92,59 @@ func (s *Store) UpdateSearch(r *router.ResolvedURL) error {
 		}
 		return errgo.Notef(err, "cannot get %s", r)
 	}
-	baseEntity, err := s.FindBaseEntity(entity.BaseURL, "acls")
+	baseEntity, err := s.FindBaseEntity(entity.BaseURL)
 	if err != nil {
-		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+		return errgo.Notef(err, "cannot get %s", entity.BaseURL)
 	}
-	doc, err := s.searchDocFromEntity(&entity, baseEntity)
+	if err := s.updateSearchEntity(&entity, baseEntity); err != nil {
+		return errgo.Notef(err, "cannot update search record for %q", entity.URL)
+	}
+	return nil
+}
+
+// UpdateSearchBaseURL updates the search record for all entities with the specified base URL.
+func (s *Store) UpdateSearchBaseURL(baseURL *charm.Reference) error {
+	if s.ES == nil || s.ES.Database == nil {
+		return nil
+	}
+	if baseURL.Series != "" {
+		return errgo.New("base url cannot contain series")
+	}
+	if baseURL.Revision != -1 {
+		return errgo.New("base url cannot contain revision")
+	}
+	// Note: It it possible to return the complete entity here and save some
+	// database round trips. Unfortunately the version of mongoDB we support
+	// (2.4) would require every field to be enumerated in this query, which
+	// would make it too fragile.
+	iter := s.DB.Entities().Pipe([]bson.D{
+		{{"$match", bson.D{{"baseurl", baseURL}}}},
+		{{"$sort", bson.D{{"revision", 1}}}},
+		{{"$group", bson.D{
+			{"_id", "$series"},
+			{"url", bson.D{{"$last", "$_id"}}},
+		}}},
```

Comment:
Should this be part of /charmstore/store.go as in FindLatestEntitesForBaseEntity? It would be nice if we have all direct access to DB in one place instead of all over the service. But I'm aware that we don't follow this rule already.